graveler/committed/tree/writer.go

```diff
@@ -0,0 +1,180 @@
+package tree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/treeverse/lakefs/graveler"
+	"github.com/treeverse/lakefs/graveler/committed"
+)
+
+type writer struct {
+	manager    committed.PartManager
+	partWriter committed.Writer // writer for the current part
+	partPath   committed.Namespace
+	lastKey    committed.Key
+	treeWriter committed.Writer // writer for the tree representation
+	treePath   committed.Namespace
+	breakRatio int // indicates when to break the parts
+}
+
+var ErrUnsorted = errors.New("record should be bigger then all records previously written to writer")
+var ErrNilValue = errors.New("record value should not be nil")
+
+func NewWriter(manager committed.PartManager, breakRatio int, treePath, partPath committed.Namespace) Writer {
+	return &writer{
+		manager:    manager,
+		breakRatio: breakRatio,
+		partPath:   partPath,
+		treePath:   treePath,
+	}
+}
+
+// WriteRecord writes a record to the current part, decides if should close part
+func (w *writer) WriteRecord(record graveler.ValueRecord) error {
+	if w.lastKey != nil && bytes.Compare(record.Key, w.lastKey) < 0 {
+		return ErrUnsorted
+	}
+	if record.Value == nil {
+		return ErrNilValue
+	}
+
+	var err error
+	if w.partWriter == nil {
+		w.partWriter, err = w.manager.GetWriter(w.partPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	v, err := MarshalValue(record.Value)
+	if err != nil {
+		return err
+	}
+	err = w.partWriter.WriteRecord(committed.Record{Key: committed.Key(record.Key), Value: v})
+	if err != nil {
+		return err
+	}
+	w.lastKey = committed.Key(record.Key)
+	breakPoint, err := w.breakPoint(record.Key)
+	if err != nil {
+		return err
+	}
+	if breakPoint {
+		if err = w.closeCurrentPart(true); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *writer) closeCurrentPart(reachedBrakePoint bool) error {
+	if w.partWriter == nil {
+		return nil
+	}
+	wr, err := w.partWriter.Close()
```

Comment:
Notice this simple `Close()` call stores the file to s3. During the write of the tree you currently serially wait for every one of those inner Stores To end!
Look for `BatchWriterCloser` that was designed specifically to handle this problem.